src/Compilers/VisualBasic/Test/Emit/Attributes/AttributeTests.vb

```diff
@@ -4337,5 +4337,160 @@ BC30002: Type 'xyz' is not defined.
+End Namespace
+Namespace TestReference
+    <Microsoft.CodeAnalysis.Embedded>
+    Public Class TestType1
+    End Class
+    <Microsoft.CodeAnalysis.EmbeddedAttribute>
+    Public Class TestType2
+    End Class
+    Public Class TestType3
+    End Class
+End Namespace
+]]>
+    </file>
+</compilation>
+
+            Dim referenceCompilation = CreateCompilationWithMscorlib(reference).ToMetadataReference()
+
+            Dim code = "
+Public Class Program
+    Public Shared Sub Main()
+        Dim obj1 = New TestReference.TestType1()
+        Dim obj2 = New TestReference.TestType2()
+        Dim obj3 = New TestReference.TestType3() ' This should be fine
+    End Sub
+End Class"
+
+            Dim compilation = CreateCompilationWithMscorlib(code, references:={referenceCompilation})
+
+            AssertTheseDiagnostics(compilation, <![CDATA[
+BC30002: Type 'TestReference.TestType1' is not defined.
+        Dim obj1 = New TestReference.TestType1()
+                       ~~~~~~~~~~~~~~~~~~~~~~~
+BC30002: Type 'TestReference.TestType2' is not defined.
+        Dim obj2 = New TestReference.TestType2()
+                       ~~~~~~~~~~~~~~~~~~~~~~~
+]]>)
+        End Sub
+
+        <Fact>
+        Public Sub EmbeddedAttributeInSourceIsAllowedIfCompilerDoesNotNeedToGenerateOne()
+
+            Dim compilation = CreateCompilationWithMscorlib(options:=TestOptions.ReleaseExe, sources:=
+<compilation>
+    <file name="a.vb"><![CDATA[
+Namespace Microsoft.CodeAnalysis
+    Friend Class EmbeddedAttribute
+        Inherits System.Attribute
+    End Class
+End Namespace
+Namespace OtherNamespace
+    <Microsoft.CodeAnalysis.Embedded>
+    Public Class TestReference
+        Public Shared Function GetValue() As Integer
+            Return 3
+        End Function
+    End Class
+End Namespace
+Public Class Program
+    Public Shared Sub Main()
+        ' This should be fine, as the compiler doesn't need to use an embedded attribute for this compilation
+        System.Console.Write(OtherNamespace.TestReference.GetValue())
+    End Sub
+End Class
+]]>
+    </file>
+</compilation>)
+
+            CompileAndVerify(compilation, expectedOutput:="3")
+        End Sub
+
+        <Fact>
+        Public Sub EmbeddedTypesInAnAssemblyAreNotExposedExternally()
+
+            Dim reference = CreateCompilationWithMscorlib(options:=TestOptions.ReleaseDll, sources:=
+<compilation>
+    <file name="a.vb"><![CDATA[
+Namespace Microsoft.CodeAnalysis
+    Friend Class EmbeddedAttribute
+        Inherits System.Attribute
+    End Class
+End Namespace
+<Microsoft.CodeAnalysis.Embedded>
+Public Class TestReference1
+End Class
+Public Class TestReference2
+End Class
+]]>
+    </file>
+</compilation>).EmitToImageReference()
+
+            Dim compilation = CreateCompilationWithMscorlib("", references:={reference})
+
+            Assert.Null(compilation.GetTypeByMetadataName("TestReference1"))
+            Assert.NotNull(compilation.GetTypeByMetadataName("TestReference2"))
```

Comment:
In first compilation, it will be exposed. Lookup will succeed (because of comparing modules in the code below). Will add the assert anyways.

```vb
If unwrappedSym.ContainingModule IsNot Me.ContainingModule Then
                If unwrappedSym.IsHiddenByVisualBasicEmbeddedAttribute() OrElse unwrappedSym.IsHiddenByCodeAnalysisEmbeddedAttribute() Then
                    Return SingleLookupResult.Empty
                End If
            End If
```